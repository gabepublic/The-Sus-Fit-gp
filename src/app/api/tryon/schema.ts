import { z } from 'zod';

/**
 * Zod schema for validating try-on request payload
 * 
 * Validates that:
 * - modelImage is a non-empty string (base64 encoded image)
 * - apparelImages is an array of non-empty strings (base64 encoded images)
 * - At least one apparel image is provided
 */
export const TryonSchema = z.object({
  modelImage: z.string().min(1, 'Missing model image'),
  apparelImages: z.array(z.string().min(1)).min(1, 'At least one apparel image required')
});

/**
 * TypeScript type inferred from the Zod schema
 */
export type TryonRequest = z.infer<typeof TryonSchema>; 