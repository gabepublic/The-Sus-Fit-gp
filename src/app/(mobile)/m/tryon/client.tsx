'use client';

import { useState, useEffect, useCallback } from 'react';
import { useRouter } from 'next/navigation';
import { TryItOnWithErrorBoundary, TryItOnContainer } from '@/mobile/components/TryItOn';
import { RouteGuard } from '@/mobile/components/RouteGuard';

interface MobileTryonClientProps {
  searchParams?: { [key: string]: string | string[] | undefined };
}

export default function MobileTryonClient({ searchParams: _searchParams }: MobileTryonClientProps) {
  const router = useRouter();
  const [hasUnsavedChanges, setHasUnsavedChanges] = useState(false);
  const [isProcessing, setIsProcessing] = useState(false);
  const [hasGeneratedResult, setHasGeneratedResult] = useState(false);

  /**
   * Browser history and back button handling
   */
  useEffect(() => {
    // Add custom browser history state for this page
    const currentState = {
      page: 'tryon',
      hasUnsavedChanges: false,
      timestamp: Date.now()
    };

    // Replace current history state to include our metadata
    if (typeof window !== 'undefined' && window.history.replaceState) {
      window.history.replaceState(currentState, '', window.location.href);
    }

    // Handle browser back button
    const handlePopState = (_event: PopStateEvent) => {
      // If user has unsaved changes, prevent navigation and show confirmation
      if (hasUnsavedChanges && (isProcessing || hasGeneratedResult)) {
        const shouldLeave = window.confirm(
          isProcessing
            ? "Your try-on is still processing. Leaving this page will cancel the process. Are you sure?"
            : "You have a try-on result that hasn't been shared yet. Are you sure you want to leave?"
        );

        if (!shouldLeave) {
          // Push the current state back to prevent navigation
          window.history.pushState(currentState, '', window.location.href);
          return;
        }
      }

      // Allow navigation to continue
      console.log('Browser back navigation allowed');
    };

    // Handle beforeunload for page refresh/close
    const handleBeforeUnload = (event: BeforeUnloadEvent) => {
      if (hasUnsavedChanges && (isProcessing || hasGeneratedResult)) {
        const message = isProcessing
          ? "Your try-on is still processing. Leaving this page will cancel the process."
          : "You have a try-on result that hasn't been shared yet.";

        event.preventDefault();
        event.returnValue = message;
        return message;
      }
    };

    // Add event listeners
    window.addEventListener('popstate', handlePopState);
    window.addEventListener('beforeunload', handleBeforeUnload);

    // Cleanup
    return () => {
      window.removeEventListener('popstate', handlePopState);
      window.removeEventListener('beforeunload', handleBeforeUnload);
    };
  }, [hasUnsavedChanges, isProcessing, hasGeneratedResult]);

  /**
   * Update browser history state when unsaved changes occur
   */
  useEffect(() => {
    if (typeof window !== 'undefined' && window.history.replaceState) {
      const updatedState = {
        page: 'tryon',
        hasUnsavedChanges,
        isProcessing,
        hasGeneratedResult,
        timestamp: Date.now()
      };

      window.history.replaceState(updatedState, '', window.location.href);
    }
  }, [hasUnsavedChanges, isProcessing, hasGeneratedResult]);

  /**
   * Access condition - user should have uploaded photos to reach try-on
   */
  const checkTryonAccess = useCallback(async (): Promise<boolean> => {
    // In a real app, this would check if user has completed upload-fit
    // For now, we'll allow access but could add conditions like:
    // - Check for uploaded photos in session storage
    // - Verify user has completed previous steps
    // - Check authentication state

    // Basic check for development - allow access
    return true;
  }, []);

  /**
   * Handle when try-on starts processing
   */
  const _handleTryItOnStart = useCallback(() => {
    setIsProcessing(true);
    setHasUnsavedChanges(true);
  }, []);

  /**
   * Handle when try-on completes successfully
   */
  const handleTryItOnComplete = useCallback((generatedImageUrl: string) => {
    setIsProcessing(false);
    setHasGeneratedResult(true);
    setHasUnsavedChanges(true); // User has a result they might want to share
    console.log('Try-on completed with result:', generatedImageUrl);
  }, []);

  /**
   * Handle try-on errors
   */
  const handleTryItOnError = useCallback((error: Error | unknown) => {
    setIsProcessing(false);
    console.error('Try-on error:', error);
  }, []);

  /**
   * Handle when user navigates to share (completed workflow)
   */
  const _handleNavigateToShare = useCallback((generatedImageUrl: string) => {
    setHasUnsavedChanges(false); // User is proceeding with the workflow
    console.log('Navigating to share with image:', generatedImageUrl);
  }, []);

  /**
   * Handle when navigation is blocked
   */
  const handleNavigationBlocked = useCallback(() => {
    console.log('Navigation blocked - user has unsaved try-on results');
  }, []);

  /**
   * Handle access denied - redirect to upload-fit
   */
  const handleAccessDenied = useCallback(() => {
    console.log('Access denied to try-on page - redirecting to upload-fit');
    router.push('/m/upload-fit');
  }, [router]);

  /**
   * Handle navigation to share - wrapper for container's onNavigate
   */
  const handleContainerNavigate = useCallback((route: string) => {
    // Set state synchronously before navigation to prevent popstate interference
    setHasUnsavedChanges(false); // User is proceeding with the workflow

    // Use setTimeout to ensure state update is processed before navigation
    setTimeout(() => {
      router.push(route);
    }, 0);
  }, [router]);

  /**
   * Mobile layout styling effects
   */
  useEffect(() => {
    // Store original backgrounds for cleanup
    const originalBodyBackground = document.body.style.backgroundColor || '';
    const originalBodyBackgroundImage = document.body.style.backgroundImage || '';
    const mobileLayout = document.querySelector('.mobile-layout') as HTMLElement;
    const mobileHeader = document.querySelector('.mobile-header') as HTMLElement;
    const originalLayoutBackground = mobileLayout?.style.backgroundColor || '';
    const originalHeaderBackground = mobileHeader?.style.backgroundColor || '';
    const originalHeaderBorder = mobileHeader?.style.borderBottom || '';

    // Set try-on specific background styling - match Upload views
    document.body.style.backgroundColor = '#faba01'; // Yellow background to match Upload views
    document.body.style.backgroundImage = 'none';

    if (mobileLayout) {
      mobileLayout.style.backgroundColor = 'transparent';
    }

    if (mobileHeader) {
      mobileHeader.style.backgroundColor = 'transparent';
      mobileHeader.style.backdropFilter = 'none';
      mobileHeader.style.borderBottom = 'none';
    }

    // Cleanup on unmount
    return () => {
      document.body.style.backgroundColor = originalBodyBackground;
      document.body.style.backgroundImage = originalBodyBackgroundImage;
      if (mobileLayout) {
        mobileLayout.style.backgroundColor = originalLayoutBackground;
      }
      if (mobileHeader) {
        mobileHeader.style.backgroundColor = originalHeaderBackground;
        mobileHeader.style.backdropFilter = '';
        mobileHeader.style.borderBottom = originalHeaderBorder;
      }
    };
  }, []);

  return (
    <RouteGuard
      hasUnsavedChanges={hasUnsavedChanges && (isProcessing || hasGeneratedResult)}
      enableNavigationGuard={true}
      confirmationMessage={
        isProcessing
          ? "Your try-on is still processing. Leaving this page will cancel the process. Are you sure?"
          : hasGeneratedResult
            ? "You have a try-on result that hasn't been shared yet. Are you sure you want to leave?"
            : "Are you sure you want to leave this page?"
      }
      protected={true}
      accessCondition={checkTryonAccess}
      onNavigationBlocked={handleNavigationBlocked}
      onAccessDenied={handleAccessDenied}
      loadingFallback={
        <div className="flex items-center justify-center min-h-[300px] text-gray-500">
          <div className="text-center">
            <div className="w-12 h-12 border-4 border-blue-200 border-t-blue-500 rounded-full animate-spin mx-auto mb-4"></div>
            <p>Initializing try-on interface...</p>
          </div>
        </div>
      }
      accessDeniedFallback={
        <div className="flex flex-col items-center justify-center min-h-[400px] p-6 text-center">
          <div className="mb-4">
            <svg
              width="64"
              height="64"
              viewBox="0 0 24 24"
              fill="none"
              stroke="currentColor"
              strokeWidth="2"
              className="text-orange-500"
            >
              <path d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"/>
              <line x1="12" y1="9" x2="12" y2="13"/>
              <line x1="12" y1="17" x2="12.01" y2="17"/>
            </svg>
          </div>
          <h2 className="text-xl font-semibold text-gray-900 mb-2">Upload Required</h2>
          <p className="text-gray-600 mb-4">
            You need to upload your photos before using the try-on feature.
          </p>
          <button
            onClick={() => router.push('/m/upload-fit')}
            className="px-6 py-3 bg-pink-500 text-white font-semibold rounded-lg hover:bg-pink-600 transition-colors"
          >
            Upload Photos
          </button>
        </div>
      }
    >
      <div className="mobile-tryon-page">
        <TryItOnWithErrorBoundary
          testId="mobile-tryon-page"
          className="tryon-interface"
          onError={(error, _errorInfo) => {
            handleTryItOnError(error);

            // Log error for monitoring in production
            console.error('Try-on page error:', error);
            if (process.env.NODE_ENV === 'production') {
              // Send to error reporting service
              // errorReportingService.captureException(error, {
              //   tags: { page: 'tryon', platform: 'mobile' },
              //   extra: { componentStack: errorInfo.componentStack, searchParams }
              // });
            }
          }}
        >
          <TryItOnContainer
            testId="mobile-tryon-container"
            className="tryon-container"
            onNavigate={handleContainerNavigate}
            onSuccess={handleTryItOnComplete}
            onError={handleTryItOnError}
          />
        </TryItOnWithErrorBoundary>
      </div>
    </RouteGuard>
  );
}