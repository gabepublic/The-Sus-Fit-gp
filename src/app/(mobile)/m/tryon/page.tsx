import { Metadata } from 'next'

export const metadata: Metadata = {
  title: 'Try It On - The Sus Fit Mobile',
  description: 'Virtual try-on results on mobile',
}

interface MobileTryonPageProps {
  searchParams?: { [key: string]: string | string[] | undefined }
}

export default function MobileTryonPage({ searchParams }: MobileTryonPageProps) {
  return (
    <div className="mobile-tryon-page">
      <h1>Try It On</h1>
      <p>Mobile Try-on Page - See how the garment looks on you</p>
      {/* Placeholder content - actual implementation will be added in future tasks */}
      <div className="placeholder-content">
        <p>Virtual try-on results will appear here</p>
        <p>See how the selected garment looks on your photo</p>
      </div>
    </div>
  )
}