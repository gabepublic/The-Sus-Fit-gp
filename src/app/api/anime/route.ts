import { NextRequest, NextResponse } from 'next/server';
import { validateBase64Image, generateAnime } from '@/lib/animeUtils';

// Types for request/response
interface AnimeRequest {
  selfie: string;
}

interface AnimeResponse {
  image: string;
}

/**
 * POST handler for anime transformation
 */
export async function POST(request: NextRequest): Promise<NextResponse> {
  try {
    // Parse request body
    const body: AnimeRequest = await request.json();
    
    // Validate input
    if (!body.selfie) {
      return NextResponse.json(
        { error: 'Missing selfie parameter' },
        { status: 400 }
      );
    }
    
    if (!validateBase64Image(body.selfie)) {
      return NextResponse.json(
        { error: 'Invalid base64 image format' },
        { status: 400 }
      );
    }
    
    // Generate anime portrait
    const animeImage = await generateAnime(body.selfie);
    
    // Return success response
    const response: AnimeResponse = {
      image: animeImage
    };
    
    return NextResponse.json(response);
    
  } catch (error) {
    console.error('Anime transformation error:', error instanceof Error ? error.message : 'Unknown error');
    
    // Return appropriate error response
    if (error instanceof Error && error.message.includes('API key')) {
      return NextResponse.json(
        { error: 'OpenAI configuration error' },
        { status: 500 }
      );
    }
    
    return NextResponse.json(
      { error: 'Failed to generate anime portrait' },
      { status: 500 }
    );
  }
} 