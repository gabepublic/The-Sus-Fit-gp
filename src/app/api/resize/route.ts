import { NextRequest, NextResponse } from 'next/server';
import { imageResizingService, ResizeOptions } from '@/lib/imageResizingService';

export async function POST(request: NextRequest) {
  try {
    const body = await request.json();
    const { imageB64, options } = body;

    console.log('route: imageB64', imageB64.substring(0, 50) + '...');
    console.log('route: options', options);
    let result;

    if (options) {
      // Resize using custom options
      const resizeOptions: ResizeOptions = {
        width: options.width,
        height: options.height,
        fit: options.fit || 'cover',
        quality: options.quality || 80,
        format: options.format || 'jpeg',
        withoutEnlargement: options.withoutEnlargement !== false,
      };
      result = await imageResizingService.resizeImage(imageB64, resizeOptions);
    } else {
      return NextResponse.json(
        { error: 'Options must be provided' },
        { status: 400 }
      );
    }

    if (!result.success) {
      return NextResponse.json(
        { 
          error: 'Image resizing failed',
          details: result.error 
        },
        { status: 500 }
      );
    }


    // Return the resized image information
    return NextResponse.json({
      success: true,
      message: 'Image resized successfully',
      resizedB64: result.resizedB64,
      metadata: result.metadata,
      resizeInfo: result.resizeInfo,
      },
      { status: 200 });

  } catch (error) {
    console.error('Image resizing error:', error);
    
    return NextResponse.json(
      { 
        error: 'Internal server error during image resizing',
        details: process.env.NODE_ENV === 'development' ? error : undefined
      },
      { status: 500 }
    );
  }
}

