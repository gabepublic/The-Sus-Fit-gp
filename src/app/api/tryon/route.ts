import { NextRequest, NextResponse } from 'next/server';
import { ZodError } from 'zod';
import { TryonSchema } from './schema';
import { getEnv } from '../../../lib/getEnv';
import { generateTryOn as generateTryOnOpenAI } from '../../../lib/openaiClient';
import { generateTryOn as generateTryOnGoogle } from '../../../lib/googleClient';
import type { TryOnParams } from '../../../lib/tryOnSchema';

import { RateLimiterMemory } from "rate-limiter-flexible";

const { dailyLimit, provider } = getEnv();
console.log('Daily Limit:', dailyLimit);

const rateLimiter = new RateLimiterMemory({
  points: dailyLimit, // 5 requests
  duration: 86400, // per 24 hours in seconds per IP
});

// TODO: Implement try-on functionality
// - ✅ Add request validation (subtask 3.2) - COMPLETED
// - ✅ Add request parsing & payload validation (subtask 3.3) - COMPLETED
// - ✅ Add image processing logic (subtask 3.4) - COMPLETED
// - ✅ Add error handling (subtask 3.5) - COMPLETED
// - ✅ Add dynamic provider switching (subtask 3.6) - COMPLETED

/**
 * Factory function to get the appropriate generateTryOn implementation
 * based on the VISION_PROVIDER environment variable
 * 
 * @returns The generateTryOn function for the configured provider
 * @throws Error if the provider is not supported
 */
function getGenerateTryOnImplementation(): (params: TryOnParams) => Promise<string> {
  //const { provider } = getEnv();
  console.log('Provider:', provider);
  switch (provider) {
    case 'OPENAI':
      return generateTryOnOpenAI;
    case 'GOOGLE':
      return generateTryOnGoogle;
    default:
      throw new Error(`Unsupported vision provider: ${provider}. Supported providers: OPENAI, GOOGLE`);
  }
}

// Helper function to create CORS headers
function createCorsHeaders(origin: string) {
  return {
    'Access-Control-Allow-Origin': origin,
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type'
  };
}

// Handle OPTIONS preflight requests
export async function OPTIONS(request: NextRequest): Promise<NextResponse> {
  const origin = request.headers.get('origin') ?? process.env.NEXT_PUBLIC_BASE_URL ?? '';
  return new NextResponse(null, {
    status: 200,
    headers: createCorsHeaders(origin)
  });
}

export async function POST(request: NextRequest): Promise<NextResponse> {
  const origin = request.headers.get('origin') ?? process.env.NEXT_PUBLIC_BASE_URL ?? '';
  const corsHeaders = createCorsHeaders(origin);

  try {
    const ip = request.headers.get("x-forwarded-for") || request.headers.get("x-real-ip") || "unknown";

    await rateLimiter.consume(ip.toString()); // throws if limit exceeded

    // Parse request body
    const body = await request.json();
    
    // Validate payload against schema
    const { modelImage, apparelImages } = TryonSchema.parse(body);
    
    // Get the appropriate generateTryOn implementation based on VISION_PROVIDER
    const generateTryOn = getGenerateTryOnImplementation();
    
    // Generate try-on image using the configured provider
    const img_generated = await generateTryOn({ modelImage, apparelImages });
    
    // Return generated image with CORS headers
    return NextResponse.json({ 
      img_generated 
    }, { 
      status: 200,
      headers: corsHeaders
    });
    
  } catch (error) {
    // Handle Zod validation errors
    if (error instanceof ZodError) {
      return NextResponse.json({ 
        error: 'Validation failed',
        details: error.errors.map(err => ({
          field: err.path.join('.'),
          message: err.message
        }))
      }, { 
        status: 400,
        headers: corsHeaders
      });
    }
    
    // Handle generateTryOn errors and other internal errors
    console.error('Try-on API error:', error);
    const errorMessage = process.env.NODE_ENV === 'development' 
      ? (error instanceof Error ? error.message : 'Unknown error occurred')
      : 'Internal Server Error';
    
    return NextResponse.json({ 
      error: errorMessage
    }, { 
      status: 500,
      headers: corsHeaders
    });
  }
} 