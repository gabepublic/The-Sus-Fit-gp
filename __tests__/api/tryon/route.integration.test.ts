import { POST, OPTIONS } from '../../../src/app/api/tryon/route';

// Mock generateTryOn function
jest.mock('../../../src/lib/openaiClient', () => ({
  generateTryOn: jest.fn()
}));

// Mock Next.js server components
jest.mock('next/server', () => {
  const mockNextResponse = jest.fn((body, options) => ({
    status: options?.status || 200,
    headers: options?.headers || {},
    json: jest.fn().mockResolvedValue(body)
  }));
  
  (mockNextResponse as any).json = jest.fn((data, options) => ({
    status: options?.status || 200,
    json: jest.fn().mockResolvedValue(data),
    headers: options?.headers || {}
  }));
  
  return {
    NextRequest: jest.fn(),
    NextResponse: mockNextResponse
  };
});

describe('/api/tryon Integration Tests', () => {
  const { generateTryOn } = require('../../../src/lib/openaiClient');

  // Valid base64 test data (1x1 pixel PNG)
  const validBase64Image = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==';

  const createMockRequest = (body: any, origin?: string | null) => {
    return {
      json: jest.fn().mockResolvedValue(body),
      headers: {
        get: jest.fn().mockReturnValue(origin)
      }
    } as any;
  };

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset environment variables
    delete (process.env as any).NEXT_PUBLIC_BASE_URL;
    delete (process.env as any).NODE_ENV;
  });

  describe('POST /api/tryon - Success Cases', () => {
    it('should return 200 with generated image for valid payload', async () => {
      const mockGeneratedImage = 'img123';
      (generateTryOn as jest.Mock).mockResolvedValue(mockGeneratedImage);

      const validPayload = {
        modelImage: validBase64Image,
        apparelImages: [validBase64Image]
      };

      const request = createMockRequest(validPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(200);
      const responseData = await response.json();
      expect(responseData).toEqual({
        img_generated: mockGeneratedImage
      });

      expect(generateTryOn).toHaveBeenCalledTimes(1);
      expect(generateTryOn).toHaveBeenCalledWith({
        modelImage: validPayload.modelImage,
        apparelImages: validPayload.apparelImages
      });

      // Verify CORS headers are present
      expect(response.headers).toMatchObject({
        'Access-Control-Allow-Origin': 'http://localhost:3000',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      });
    });

    it('should handle multiple apparel images correctly', async () => {
      const mockGeneratedImage = 'img456';
      (generateTryOn as jest.Mock).mockResolvedValue(mockGeneratedImage);

      const validPayload = {
        modelImage: validBase64Image,
        apparelImages: [validBase64Image, validBase64Image]
      };

      const request = createMockRequest(validPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(200);
      const responseData = await response.json();
      expect(responseData).toEqual({
        img_generated: mockGeneratedImage
      });

      expect(generateTryOn).toHaveBeenCalledWith({
        modelImage: validPayload.modelImage,
        apparelImages: validPayload.apparelImages
      });
    });
  });

  describe('POST /api/tryon - Validation Error Cases', () => {
    it('should return 400 for missing apparelImages with Zod error', async () => {
      const invalidPayload = {
        modelImage: 'invalid-base64-data'
        // Missing apparelImages
      };

      const request = createMockRequest(invalidPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(400);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: 'Validation failed'
      });

      // Test Zod error message snapshot
      expect(responseData.details).toMatchSnapshot();
      expect(responseData.details).toHaveLength(2);
      expect(responseData.details[0].field).toBe('modelImage');
    });

    it('should return 400 for empty apparelImages array', async () => {
      const invalidPayload = {
        modelImage: 'invalid-base64-data',
        apparelImages: []
      };

      const request = createMockRequest(invalidPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(400);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: 'Validation failed'
      });

      expect(responseData.details).toMatchSnapshot();
      expect(responseData.details).toHaveLength(2);
      expect(responseData.details[0].field).toBe('modelImage');
    });

    it('should return 400 for missing modelImage', async () => {
      const invalidPayload = {
        apparelImages: ['invalid-base64-data']
        // Missing modelImage
      };

      const request = createMockRequest(invalidPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(400);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: 'Validation failed'
      });

      expect(responseData.details).toMatchSnapshot();
      expect(responseData.details).toHaveLength(2);
      expect(responseData.details[0].field).toBe('modelImage');
    });

    it('should return 400 for invalid base64 data', async () => {
      const invalidPayload = {
        modelImage: 'invalid-base64-data',
        apparelImages: ['invalid-base64-data']
      };

      const request = createMockRequest(invalidPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(400);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: 'Validation failed'
      });

      expect(responseData.details).toMatchSnapshot();
    });
  });

  describe('POST /api/tryon - Internal Error Cases', () => {
    it('should return 500 when generateTryOn throws an error', async () => {
      const validPayload = {
        modelImage: validBase64Image,
        apparelImages: [validBase64Image]
      };

      (generateTryOn as jest.Mock).mockRejectedValue(new Error('OpenAI API error'));

      const request = createMockRequest(validPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(500);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: 'Internal Server Error'
      });

      expect(generateTryOn).toHaveBeenCalledTimes(1);
    });

    it('should return 500 for malformed JSON', async () => {
      const request = {
        json: jest.fn().mockRejectedValue(new Error('Invalid JSON')),
        headers: {
          get: jest.fn().mockReturnValue('http://localhost:3000')
        }
      } as any;

      const response = await POST(request);

      expect(response.status).toBe(500);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: 'Internal Server Error'
      });
    });

    it('should return detailed error message in development environment', async () => {
      (process.env as any).NODE_ENV = 'development';
      
      const validPayload = {
        modelImage: validBase64Image,
        apparelImages: [validBase64Image]
      };

      const errorMessage = 'Development error message';
      (generateTryOn as jest.Mock).mockRejectedValue(new Error(errorMessage));

      const request = createMockRequest(validPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(500);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: errorMessage
      });
    });

    it('should return generic error message in production environment', async () => {
      (process.env as any).NODE_ENV = 'production';
      
      const validPayload = {
        modelImage: validBase64Image,
        apparelImages: [validBase64Image]
      };

      (generateTryOn as jest.Mock).mockRejectedValue(new Error('Production error message'));

      const request = createMockRequest(validPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(500);
      const responseData = await response.json();
      expect(responseData).toMatchObject({
        error: 'Internal Server Error'
      });
    });
  });

  describe('OPTIONS /api/tryon - CORS Preflight', () => {
    it('should return 200 for OPTIONS preflight request', async () => {
      const request = createMockRequest({}, 'http://localhost:3000');
      const response = await OPTIONS(request);

      expect(response.status).toBe(200);
      expect(response.headers).toMatchObject({
        'Access-Control-Allow-Origin': 'http://localhost:3000',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      });
    });

    it('should use NEXT_PUBLIC_BASE_URL when origin header is null', async () => {
      (process.env as any).NEXT_PUBLIC_BASE_URL = 'https://example.com';
      
      const request = createMockRequest({}, null);
      const response = await OPTIONS(request);

      expect(response.status).toBe(200);
      expect(response.headers).toMatchObject({
        'Access-Control-Allow-Origin': 'https://example.com',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      });
    });

    it('should use empty string when both origin header and NEXT_PUBLIC_BASE_URL are null/undefined', async () => {
      const request = createMockRequest({}, null);
      const response = await OPTIONS(request);

      expect(response.status).toBe(200);
      expect(response.headers).toMatchObject({
        'Access-Control-Allow-Origin': '',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      });
    });
  });

  describe('CORS Headers in POST Responses', () => {
    it('should include CORS headers in successful POST response', async () => {
      const mockGeneratedImage = 'img123';
      (generateTryOn as jest.Mock).mockResolvedValue(mockGeneratedImage);

      const validPayload = {
        modelImage: validBase64Image,
        apparelImages: [validBase64Image]
      };

      const request = createMockRequest(validPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(200);
      expect(response.headers).toMatchObject({
        'Access-Control-Allow-Origin': 'http://localhost:3000',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      });
    });

    it('should include CORS headers in error responses', async () => {
      const invalidPayload = {
        modelImage: validBase64Image
        // Missing apparelImages
      };

      const request = createMockRequest(invalidPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(400);
      expect(response.headers).toMatchObject({
        'Access-Control-Allow-Origin': 'http://localhost:3000',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      });
    });

    it('should include CORS headers in internal error responses', async () => {
      const validPayload = {
        modelImage: validBase64Image,
        apparelImages: [validBase64Image]
      };

      (generateTryOn as jest.Mock).mockRejectedValue(new Error('Test error'));

      const request = createMockRequest(validPayload, 'http://localhost:3000');
      const response = await POST(request);

      expect(response.status).toBe(500);
      expect(response.headers).toMatchObject({
        'Access-Control-Allow-Origin': 'http://localhost:3000',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      });
    });
  });
});
