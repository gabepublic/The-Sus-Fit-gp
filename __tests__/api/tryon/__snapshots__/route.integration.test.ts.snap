// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`/api/tryon Integration Tests POST /api/tryon - Validation Error Cases should return 400 for empty apparelImages array 1`] = `
[
  {
    "field": "modelImage",
    "message": "Invalid base64 image data",
  },
  {
    "field": "apparelImages",
    "message": "At least one apparel image required",
  },
]
`;

exports[`/api/tryon Integration Tests POST /api/tryon - Validation Error Cases should return 400 for invalid base64 data 1`] = `
[
  {
    "field": "modelImage",
    "message": "Invalid base64 image data",
  },
  {
    "field": "apparelImages.0",
    "message": "Invalid base64 image data",
  },
]
`;

exports[`/api/tryon Integration Tests POST /api/tryon - Validation Error Cases should return 400 for missing apparelImages with Zod error 1`] = `
[
  {
    "field": "modelImage",
    "message": "Invalid base64 image data",
  },
  {
    "field": "apparelImages",
    "message": "Required",
  },
]
`;

exports[`/api/tryon Integration Tests POST /api/tryon - Validation Error Cases should return 400 for missing modelImage 1`] = `
[
  {
    "field": "modelImage",
    "message": "Required",
  },
  {
    "field": "apparelImages.0",
    "message": "Invalid base64 image data",
  },
]
`;
